### PR TITLE
[KTO] fix minor bugs with data loading and reporting

### PR DESCRIPTION
**Pull request by kawine**
- remove interleaving in KTOTrainer since it is no longer needed for KL estimation and artificially duplicates data to force an even number of positive and negative examples (see https://github.com/huggingface/trl/issues/1467)
- do not report batch statistics for good/bad data if there are zero good/bad examples in the current batch (currently NaNs are returned, which are sometimes converted to floats before being sent to wandb, leading to reward spikes (see https://twitter.com/_lewtun/status/1771493512262942801)

cc @kashif

### CODE REVIEW

**Comment by kashif** [comment]
thanks @kawine! could you also have a quick peek at https://github.com/huggingface/trl/pull/1472

File: tests/test_kto_trainer.py [hunk 1]
```diff
@@ -39,6 +40,7 @@ def setUpClass(cls):
         cls.t5_model = AutoModelForSeq2SeqLM.from_pretrained(model_id)
         cls.t5_ref_model = AutoModelForSeq2SeqLM.from_pretrained(model_id)
         cls.t5_tokenizer = AutoTokenizer.from_pretrained(model_id)
+        wandb.init(mode="disabled")
```
Comment (lewtun) [inline tests/test_kto_trainer.py:43]: I would prefer to replace this with a decorator on the whole class like this:

```
from unittest.mock import patch

@patch.dict(os.environ, {"WANDB_DISABLED": "true"})
```

**Review by lewtun — APPROVED** [review]
Thanks for these fixes @kawine ! LGTM with a small suggestion to use a decorator for the WANDB env variable

File: trl/trainer/kto_trainer.py [hunk 2]
```diff
@@ -999,13 +989,25 @@ def get_batch_loss_metrics(
...
+        metrics[f"{prefix}rewards/chosen"] = self.accelerator.gather(mean_chosen_reward).nanmean().item()
+        metrics[f"{prefix}logps/chosen"] = self.accelerator.gather(mean_chosen_logps).nanmean().item()
+        metrics[f"{prefix}rewards/rejected"] = self.accelerator.gather(mean_rejected_reward).nanmean().item()
+        metrics[f"{prefix}logps/rejected"] = self.accelerator.gather(mean_rejected_logps).nanmean().item()
         metrics[f"{prefix}rewards/margins"] = metrics[f"{prefix}rewards/chosen"] - metrics[f"{prefix}rewards/rejected"]
         metrics[f"{prefix}kl"] = kl.item()  # has already been gathered in kto_loss
-        metrics[f"{prefix}logps/chosen"] = self.accelerator.gather(mean_chosen_logps).nanmean().cpu()
-        metrics[f"{prefix}logps/rejected"] = self.accelerator.gather(mean_rejected_logps).nanmean().cpu()
 
+        if np.isnan(metrics[f"{prefix}rewards/chosen"]):
```
Comment (claralp) [inline trl/trainer/kto_trainer.py:1000]: you can omit these checks if you use
```
        logs[key] = torch.tensor(metrics).nanmean().item()
```
 in line [1203](https://github.com/huggingface/trl/pull/1476/files#diff-f4e653023a53778d00b43a8202a76dbac88b9a790987970f232bafa93e581299R1203). This ignores the nan values when logging metrics

File: trl/trainer/kto_trainer.py [hunk 3]
```diff
@@ -360,15 +360,15 @@ def make_inputs_require_grad(module, input, output):
...
+        num_undesirable = len(train_dataset) - num_desirable
 
-        if len(desirable) != len(undesirable):
+        if num_desirable != num_undesirable:
             # The lower and upper bounds come from Eq. (8) of https://arxiv.org/abs/2402.01306
-            des_weight_lower_bound = round((len(undesirable) * self.undesirable_weight / len(desirable)) * 1, 2)
-            des_weight_upper_bound = round((len(undesirable) * self.undesirable_weight / len(desirable)) * 1.33, 2)
-            und_weight_lower_bound = round((len(desirable) * self.desirable_weight / len(undesirable)) / 1.33, 2)
-            und_weight_upper_bound = round((len(desirable) * self.desirable_weight / len(undesirable)) / 1, 2)
+            des_weight_lower_bound = round((len(num_undesirable) * self.undesirable_weight / len(num_desirable)) * 1, 2)
```
Comment (claralp) [inline trl/trainer/kto_trainer.py:368]: isn´t `num_undesirable ` the int length already? For me `len(num_undesirable)` throws an error
